lib/auth/userloginstate/generator.go

```diff
@@ -0,0 +1,175 @@
+		g.Clock = clockwork.NewRealClock()
+	}
+
+	return nil
+}
+
+// Generator will generate a user login state from a user.
+type Generator struct {
+	accessLists services.AccessListsGetter
+	roles       services.RoleGetter
+	clock       clockwork.Clock
+}
+
+// NewGenerator creates a new user login state generator.
+func NewGenerator(config GeneratorConfig) (*Generator, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Generator{
+		accessLists: config.AccessLists,
+		roles:       config.Roles,
+		clock:       config.Clock,
+	}, nil
+}
+
+// Generate will generate the user login state for the given user.
+func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginstate.UserLoginState, error) {
+	// Create a new empty user login state.
+	uls, err := userloginstate.New(
+		header.Metadata{
+			Name: user.GetName(),
+		}, userloginstate.Spec{
+			Roles:  []string{},
+			Traits: trait.Traits{},
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Generate the user login state.
+	if err := g.addAccessListsToState(ctx, user, uls); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Clean up the user login state after generating it.
+	if err := g.postProcess(ctx, uls); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return uls, nil
+}
+
+// addAccessListsToState will added the user's applicable access lists to the user login state.
+func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) error {
+	accessLists, err := g.accessLists.GetAccessLists(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Create an identity for testing membership to access lists.
+	identity := tlsca.Identity{
+		Username: user.GetName(),
+		Groups:   user.GetRoles(),
+		Traits:   user.GetTraits(),
+	}
+
+	for _, accessList := range accessLists {
+		if err := services.IsMember(identity, g.clock, accessList); err != nil {
+			continue
+		}
+
+		state.Spec.Roles = append(state.Spec.Roles, accessList.Spec.Grants.Roles...)
+
+		for k, values := range accessList.Spec.Grants.Traits {
+			state.Spec.Traits[k] = append(state.Spec.Traits[k], values...)
+		}
+	}
+
+	return nil
+}
+
+// postProcess will perform cleanup to the user login state after its generation.
+func (g *Generator) postProcess(ctx context.Context, state *userloginstate.UserLoginState) error {
+	// Deduplicate roles and traits
+	state.Spec.Roles = utils.Deduplicate(state.Spec.Roles)
+
+	for k, v := range state.Spec.Traits {
+		state.Spec.Traits[k] = utils.Deduplicate(v)
+	}
+
+	// Remove roles that don't exist in the backend so that we don't generate certs for non-existent roles.
+	// Doing so can prevent login from working properly. This could occur if access lists refer to roles that
+	// no longer exist, for example.
+	existingRoles := []string{}
+	for _, role := range state.Spec.Roles {
+		if _, err := g.roles.GetRole(ctx, role); err == nil {
```

Comment:
Can we just change the available method to `GetRoles`? I think this would be faster by virtue of fewer RPC calls. We can get away from checking for `trace.IsNotFound` err, which I find imprecise at best.